### ISSUE TITLE
akka-http request parsing parses HEAD as GET

### ISSUE THREAD

**Issue by mhamrah** [akka/akka#18020]
If I issue a:

```
curl -XHEAD -v http://localhost:8080
```

to an akka-http app, the routing logic sees this as a GET request, not a HEAD request.

**Comment by mhamrah** [comment]
I didn't see the transparent-head-request option

**Comment by magro** [comment]
I'd like to reopen this issue, because IMO for a transparent HEAD request the `request.method` should be `GET` (at least this is how play handles transparent HEAD). This is s.th. that caused issues in play, see playframework/playframework#4907.

**Comment by ktoso** [comment]
Could you rephrase @magro, what change you'd like to propose?

// cc @sirthias @jrudolph

**Comment by 2beaucoup** [comment]
@magro 
> IMO for a transparent `HEAD` request the `request.method` should be `GET

Doesn't this describe the current behaviour? See OP. ;)

**Comment by magro** [comment]
Sorry, I was wrong, GET should have been HEAD, I'll correct it above ;-)

**Comment by magro** [comment]
@ktoso For a transparent HEAD request request.method should return a HEAD HttpMethod.

**Comment by jrudolph** [comment]
Hi @magro, nice to see you here ;)

There may be different understandings about what is meant by "transparent HEAD request handling".

In spray and in akka-http, it means that application code doesn't notice it is currently handling a HEAD request. If you don't want this automatic handling you can always switch the `transparent-head-request` setting off.

Maybe that's what play needs to do if it wants to provide customized transparent HEAD request handling without piggy-backing on top of what akka-http provides. What won't work (and maybe that's what you found out?) is that two layers both try to implement transparent HEAD requests at the same time.

**Comment by magro** [comment]
Hey @jrudolph, the pleasure is on my side :-)

If akka-http sets `request.method` intentionally to `GET` for transparent HEAD handling, then play should indeed set `akka.http.server.transparent-head-requests=false` - that's what I also suggested  in playframework/playframework#4907.

Thanks for the confirmation!

**Comment by magro** [comment]
@jrudolph What was the reasoning behind the decision to set request.method to GET for transparent HEAD requests?

As a user using a monitoring tool like e.g. [kamon](https://github.com/kamon-io/Kamon) that reports requests (based on request method and path), I'd expect to see different traces for HEAD and GET requests. If there are lots of HEAD requests while I have no HEAD directive defined and GETs are expensive, I could decide to add an explicit mapping/directive for HEAD.
That's the use case where I'm coming from, because we had seen unexpected behaviour / monitoring traces for transparent HEAD requests in play. This is described in kamon-io/Kamon#237.

**Comment by sirthias** [comment]
@magro The reasoning behind this setting is that, when you don't switch it off, HEAD request will automatically dealt with without you having to do *anything* in your routing layer. Your code simply sees GET requests and does what it needs to do to complete them.
The HTTP layer determines whether the full response entity needs to be send (in the case of real GET requests) or not (in the case of a "fake" GET request which really is a HEAD request).

Of course you are right and it might be that some work is done that could be avoided when HEAD requests were handled explicitly by the application. However, in many cases this additional overhead doesn't really matter and people happily choose the convenience over the extra cycles.

Note that with akka-http's streaming nature the extra work is in many cases less than it was with spray, since the response entity stream can be constructed fully lazily (i.e. only when the response entity stream is actually materialized), which the HTTP layer will not do in the case of a transparent HEAD request.

**Comment by magro** [comment]
@sirthias I agree, that for a transparent HEAD request the user shouldn't have to do anything in his routing layer. But - perhaps because play does it like this - I'd expect to see which HTTP method was actually used for a request (e.g. to be able to have correct reporting/monitoring). So I'd expect that the `get` directive/route would be called (or some that includes `get`), but with `request.method` = `HEAD`. Why shouldn't a user see only a GET request, and not that it was actually HEAD?
I don't want to criticize this decision, I just want to understand the thinking behind it, because in play this is obviously different.

Regarding costs/optimization: AFAICS the transparent HEAD request returns the Content-Length (I tried this [minimal-scala-akka-http-seed template](https://www.typesafe.com/activator/template/minimal-scala-akka-http-seed)), therefore I assume that the response body is fully constructed to determine this (and backend calls that are needed for the response body will be executed, which I'd be more concerned about than the costs for sending the response body). But that's not the point I'd want to discuss, I'm more interested in the thinking about user expectations regarding the request method.

**Comment by sirthias** [comment]
@magro in Akka HTTP (and in spray) the low-level HTTP layer and the high-level routing layer are fully decoupled. It is completely normal for an application to only use the low-level layer (and implement its own "routing").
The "transparent head requests" setting is a feature of the *low-level* layer. As such there is no direct connection to the routing DSL. But you can very easily implement a "transparent head requests" feature with the routing DSL yourself, if you so want. (Of course you should disable the `transparent-head-requests` heading for the low-level layer for this to work.)

> AFAICS the transparent HEAD request returns the Content-Length (I tried this minimal-scala-akka-http-seed template), therefore I assume that the response body is fully constructed to determine this

This depends entirely on the application. If the Content-Length is easily known without constructing the entity (as for example in the case of file content) there is no need to load the entity before it is actually being sent. And if the Content-Length is not easily available the application could also decide to return a chunked entity. This too would make it easy to not load any entity content unnecessarily for HEAD requests.

In general we are not saying that the `transparent-head-requests` setting is a perfect solution. It definitely has its weaknesses. But, since it has served us quite well over the last years with spray, we migrated it over to Akka HTTP.

**Comment by magro** [comment]
@sirthias Thanks for this explanation!

**Comment by magro** [comment]
From my point of view this can be closed again, thanks for the discussion! :-)
